core/src/main/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoder.java

```diff
@@ -0,0 +1,94 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.client.lexicoder;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.accumulo.core.clientImpl.lexicoder.ByteUtils.concat;
+import static org.apache.accumulo.core.clientImpl.lexicoder.ByteUtils.escape;
+import static org.apache.accumulo.core.clientImpl.lexicoder.ByteUtils.split;
+import static org.apache.accumulo.core.clientImpl.lexicoder.ByteUtils.unescape;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.accumulo.core.clientImpl.lexicoder.AbstractLexicoder;
+
+/**
+ * A Lexicoder to encode/decode a Java List to/from a byte array where the concatenation of each
+ * encoded element sorts lexicographically.
+ *
+ * Note: Unlike {@link ListLexicoder}, this implementation supports empty lists.
+ *
+ * The lists are encoded with the elements separated by null (0x0) bytes, which null bytes appearing
+ * in the elements escaped as two 0x1 bytes, and 0x1 bytes appearing in the elements escaped as 0x1
+ * and 0x2 bytes. The list is terminated with a final delimiter, with no bytes following it.
+ *
+ * @since 2.0.0
+ * @param <E>
+ *          list element type.
+ */
+public class SequenceLexicoder<E> extends AbstractLexicoder<List<E>> {
+
+  private static final byte[] EMPTY_ELEMENT = new byte[0];
+  private final Lexicoder<E> elementLexicoder;
+
+  /**
+   * Primary constructor.
+   *
+   * @param elementLexicoder
+   *          Lexicoder to apply to elements.
+   */
+  public SequenceLexicoder(final Lexicoder<E> elementLexicoder) {
+    this.elementLexicoder = requireNonNull(elementLexicoder, "elementLexicoder");
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @return a byte array containing the concatenation of each element in the list encoded.
+   */
+  @Override
+  public byte[] encode(final List<E> v) {
+    final byte[][] encElements = new byte[v.size() + 1][];
+    int index = 0;
+    for (final E element : v) {
+      encElements[index++] = escape(elementLexicoder.encode(element));
+    }
+    encElements[v.size()] = EMPTY_ELEMENT;
+    return concat(encElements);
+  }
+
+  @Override
+  protected List<E> decodeUnchecked(final byte[] b, final int offset, final int len) {
+    final byte[][] escapedElements = split(b, offset, len);
+    assert escapedElements.length
```

Comment:
The assertion is checking (and documenting) a code assumption, not validating arguments. If assertions are disabled, that's fine, because this is only checking that something that should never happen doesn't happen.

Reply:
Also note that if the assumption is invalid and assertions are disabled, an `ArrayIndexOutOfBoundsException` will be thrown on the next line.

Reply:
The decode method is effectively validating arbitrary user input (since it is the responsibility of the user's client code to ensure that what is passed to the decode method is something which was written after being emitted from the corresponding encode method).

Since this is validating arbitrary user input at a public API boundary, rather than asserting internal program correctness, an explicit check of the input is more appropriate than an assert statement here.

Also, `IllegalArgumentException` would be more appropriate than `ArrayIndexOutOfBoundsException` because the latter exposes, and is dependent upon, implementation details. `IllegalArgumentExcept`, on the other hand, is a helpful API exception for the caller that explicitly says there's a problem with the argument they provided.

Reply:
@reggert sorry for not getting back to this, it fell off my radar.

>  because this is only checking that something that should never happen doesn't happen.

Is this statement based on the assumption that users will only decode data that was encoded with this lexicoder?  Is the following use case possible?

  * User uses this lexicoder to encode data and store it in Accumulo column family.
  * User accidentally passes Accumulo row to this code for decoding and it triggers.  Could it throw the array out bounds exception in this case?

If the above is possible a user facing error message would be a nice touch to an already really nice change.  Maybe a message like `Either this data was corrupted or not encoded with "+this.getClass().getName()`.

Reply:
The assertion isn't checking user input. It's checking the return value of `ByteUtils.split`, and verifying that it behaves as expected.

Reply:
> The assertion isn't checking user input. It's checking the return value of ByteUtils.split, and verifying that it behaves as expected.

I see.  I misunderstood what that code was trying to do.